### DISCUSSION TITLE
Difference between BackgroundTask, Middlewares and After Request hook

### DISCUSSION THREAD

**Discussion by IntrepidCrow** [litestar-org/litestar#666] [Q&A] [answered]
Hello, everyone.

I am a novice web developer and am currently working on a Starlite-based pet project.

I don't really understand the difference between BackgroundTask, Middleware, and After Request hook. It feels like these concepts duplicate behavior, but that's probably because I don't understand them very well. Can someone explain the difference to me?

Thanks a lot in advance!

**Comment by Goldziher** [comment, marked as answer]
> Hello, everyone.
> 
> I am a novice web developer and am currently working on a Starlite-based pet project.
> 
> I don't really understand the difference between BackgroundTask, Middleware, and After Request hook. It feels like these concepts duplicate behavior, but that's probably because I don't understand them very well. Can someone explain the difference to me?
> 
> Thanks a lot in advance!

Hi @IntrepidCrow. 

So there is a degree of overlap between these, but there are also key differences. 

A `BackGroundTask` is simply a sync/async callable that is executed once the response is done sending. It has no access to request and and it cannot be defined on different layers of the application.

An `AfterResponseHookHandler` on the other hand receives the request instance as a parameter and you can "layer" it on the application, so in principle you can declare it even top level on the Starlite constructor for the entire application. You can of course do quite a few similar things in these two, but there are things you can do with an AfterReesponse handler that you cannot do with a background task. Think even of a simple example as logging values from the request instance.

As for Middleware - You can do the same as you can with the other two using middleware, but manipulating responses there is low level and more involved. Furthermore, granular control of which handlers are going to be affected by a middleware is also a bit more involved. 

All in all though, Starlite does not restrict you to use only one pattern - you can choose which one suits your use case best. I would suggest to consider the above in making this determination.
